File: list-1/ex-1.hs
{-# LANGUAGE PatternSynonyms, ViewPatterns #-}
{-# LANGUAGE DeriveGeneric #-}

module Main where

import GHC.Generics (Generic)
import Data.Hashable
import Data.HashMap.Strict as Map
import Data.List as List
import Data.Sequence as Seq

import System.IO.Unsafe

type Pos = (Int, Int)

data Color = White | Black
  deriving (Eq, Ord, Generic, Show)

instance Hashable Color

data State = State 
  { whiteTower :: Pos
  , whiteKing  :: Pos
  , blackKing  :: Pos 
  }
  deriving (Eq, Ord, Generic, Show)

instance Hashable State

next :: Color -> Color
next Black = White
next White = Black

moves :: Color -> State -> [State]
moves Black (State wt wk bk) = List.filter isValid $ fmap (\k -> State wt wk k) $ kingNeighbours bk
moves White (State wt wk bk) = List.filter isValid $ fmap (\k -> State wt k bk) (kingNeighbours wk) ++
  fmap (\t -> State t wk bk) (towerNeighbours wt)

check :: State -> Bool
check (State (tx, ty) _ (kx , ky)) = tx == kx || ty == ky

checkMate :: State -> Bool
checkMate s = all check $ moves Black s

isValid :: State -> Bool
isValid (State wt wk bk) = wt /= wk && wt /= bk && wk /= bk && all (/= wk) (kingNeighbours bk)

inRange :: Int -> Bool
inRange x = x >= 1 && x <= 8

kingNeighbours :: Pos -> [Pos]
kingNeighbours (x, y) = [(a, b) | a <- [x - 1 .. x + 1],
                                  b <- [y - 1 .. y + 1],
                                  inRange a, inRange b]

towerNeighbours :: Pos -> [Pos]
towerNeighbours (x, y) = [(x, z) | z <- [1 .. 8]] ++ [(z, y) | z <- [1 .. 8]]

strToPos :: String -> Pos
strToPos [x, y] = ((fromEnum x) - (fromEnum 'a') + 1, read [y])
strToPos _ = error "Unexpected input"

parseInput :: String -> (Color, State)
parseInput s =  
  let [c, wk, wt, bk] = words s
  in (if c == "black" then Black else White, State (strToPos wt) (strToPos wk) (strToPos bk))

choose :: [State] -> [State] -> [State]
choose s1 s2 = if List.length s1 < List.length s2 then s1 else s2

run :: Seq (Color, [State]) -> HashMap (Color, State) [State] -> Maybe [State] -> Maybe [State]
run q visited best = case Seq.viewl q of
  EmptyL -> best
  (c, s:ss) :< q ->
    let
      ms = moves c s
      q' = List.foldl (\q s' -> if Map.member (next c, s') visited then q else q |> (next c, s':s:ss)) q ms
      v' = List.foldl (\v s' -> Map.insertWith choose (next c, s') (s':s:ss) v) visited ms
      b = if checkMate s
        then case best of
          Nothing -> Just $ s:ss
          Just s' -> Just $ choose s' (s:ss)
        else best
    in -- seq (unsafePerformIO $ print (Map.size v'))
      run q' v' b



main :: IO ()
main = putStrLn "Hello world"

File: list-1/ex1.scala
import scala.collection.mutable.{ Map, HashMap, Queue }

trait Figure {
  val x: Int
  val y: Int
  def inRange(x: Int) = x >= 1 && x <= 8
  def possiblePositions: Seq[Figure]
  def pos = (x, y)
} 

case class King(x: Int, y: Int) extends Figure {
  def possiblePositions: Seq[King] = {
    for {
      a <- (x - 1) to (x + 1)
      b <- (y - 1) to (y + 1)
      if (inRange(a) && inRange(b))
    } yield King(a, b)
  }
}

case class Tower(x: Int, y: Int) extends Figure {
  def possiblePositions: Seq[Tower] = {
    (1 to 8).map { a => Tower(a, y) } ++
    (1 to 8).map { b => Tower(x, b) }
  }
}

trait Color {
  def next: Color
}

case object Black extends Color {
  def next = White
}

case object White extends Color {
  def next = Black
}

case class State(wt: Tower, wk: King, bk: King) {
  def isValid = {
    wt.pos != wk.pos &&
    wt.pos != bk.pos &&
    wk.pos != bk.pos &&
    bk.possiblePositions.forall { _ != wk }
  }

  def check = {
    wt.x == bk.x || wt.y == bk.y
  }

  def checkMate = moves(Black).forall(_.check)

  def moves(c: Color) = c match {
    case Black => bk.possiblePositions.map { State(wt, wk, _) }.filter(_.isValid)
    case White => 
      wk.possiblePositions.map { State(wt, _, bk) }.filter(_.isValid) ++
      wt.possiblePositions.map { State(_, wk, bk) }.filter(_.isValid)
  }
}

object Main {
  def choose[A](a: Seq[A], b: Seq[A]) = if (a.size < b.size) a else b

  def run(init: State, c: Color) = {
    val visited: Map[(Color, State), Seq[State]] = HashMap.empty
    val queue: Queue[(Color, Seq[State])] = Queue((c, Seq(init)))
    var best: Option[Seq[State]] = None
    while (queue.nonEmpty) {
      val (c, states) = queue.head
      queue.dequeue()
      val moves = states.head.moves(c)
      moves.foreach { s =>
        val newStates = s +: states
        visited.get((c.next, s)) match {
          case None =>
            queue += ((c.next, newStates))
            visited += (((c.next, s), newStates))
          case Some(ss) =>
            if (ss.size > newStates.size)
              visited.update((c.next, s), newStates)
        }
      }
      best = best match {
        case None => Some(states)
        case Some(ss) => Some(choose(ss, states))
      }
    }
    best
  }

  def main(args: Array[String]): Unit = {
    println("starting")
    println(run(State(Tower(3, 8), King(3, 4), King(8, 3)), Black))
  }
}